File: cli/src/documentation_robotics/export/archimate_exporter.py
"""
ArchiMate XML exporter.
"""
from pathlib import Path
from typing import Dict, List, Set
from lxml import etree as ET
from datetime import datetime
from .export_manager import BaseExporter, ExportOptions
from ..core.element import Element


class ArchiMateExporter(BaseExporter):
    """
    Exports model to ArchiMate 3.2 XML format.

    Creates ArchiMate Exchange File that can be imported into
    ArchiMate tools like Archi, Enterprise Architect, etc.
    """

    # ArchiMate namespace
    ARCHIMATE_NS = "http://www.opengroup.org/xsd/archimate/3.0/"
    XSI_NS = "http://www.w3.org/2001/XMLSchema-instance"

    # Layer to ArchiMate element type mapping
    LAYER_TYPE_MAP = {
        # Motivation layer
        ("motivation", "stakeholder"): "Stakeholder",
        ("motivation", "driver"): "Driver",
        ("motivation", "assessment"): "Assessment",
        ("motivation", "goal"): "Goal",
        ("motivation", "outcome"): "Outcome",
        ("motivation", "principle"): "Principle",
        ("motivation", "requirement"): "Requirement",
        ("motivation", "constraint"): "Constraint",
        ("motivation", "meaning"): "Meaning",
        ("motivation", "value"): "Value",
        # Business layer
        ("business", "actor"): "BusinessActor",
        ("business", "role"): "BusinessRole",
        ("business", "collaboration"): "BusinessCollaboration",
        ("business", "interface"): "BusinessInterface",
        ("business", "process"): "BusinessProcess",
        ("business", "function"): "BusinessFunction",
        ("business", "interaction"): "BusinessInteraction",
        ("business", "event"): "BusinessEvent",
        ("business", "service"): "BusinessService",
        ("business", "object"): "BusinessObject",
        ("business", "contract"): "Contract",
        ("business", "representation"): "Representation",
        ("business", "product"): "Product",
        # Application layer
        ("application", "component"): "ApplicationComponent",
        ("application", "collaboration"): "ApplicationCollaboration",
        ("application", "interface"): "ApplicationInterface",
        ("application", "function"): "ApplicationFunction",
        ("application", "interaction"): "ApplicationInteraction",
        ("application", "event"): "ApplicationEvent",
        ("application", "service"): "ApplicationService",
        ("application", "process"): "ApplicationProcess",
        ("application", "object"): "DataObject",
        # Technology layer
        ("technology", "node"): "Node",
        ("technology", "device"): "Device",
        ("technology", "software"): "SystemSoftware",
        ("technology", "collaboration"): "TechnologyCollaboration",
        ("technology", "interface"): "TechnologyInterface",
        ("technology", "path"): "Path",
        ("technology", "network"): "CommunicationNetwork",
        ("technology", "function"): "TechnologyFunction",
        ("technology", "process"): "TechnologyProcess",
        ("technology", "interaction"): "TechnologyInteraction",
        ("technology", "event"): "TechnologyEvent",
        ("technology", "service"): "TechnologyService",
        ("technology", "artifact"): "Artifact",
    }

    # Relationship type mapping
    RELATIONSHIP_TYPE_MAP = {
        "realization": "Realization",
        "serving": "Serving",
        "access": "Access",
        "usage": "Usage",
        "composition": "Composition",
        "aggregation": "Aggregation",
        "assignment": "Assignment",
        "flow": "Flow",
        "triggering": "Triggering",
        "specialization": "Specialization",
        "association": "Association",
    }

    def export(self) -> Path:
        """
        Export to ArchiMate XML.

        Returns:
            Path to exported file
        """
        # Create XML structure
        root = self._create_root()

        # Add model metadata
        self._add_metadata(root)

        # Add elements
        elements_section = ET.SubElement(root, "elements")
        self._add_elements(elements_section)

        # Add relationships
        relationships_section = ET.SubElement(root, "relationships")
        self._add_relationships(relationships_section)

        # Add organizations (views/folders)
        organizations_section = ET.SubElement(root, "organizations")
        self._add_organizations(organizations_section)

        # Create XML tree
        tree = ET.ElementTree(root)

        # Ensure output directory exists
        self.options.output_path.parent.mkdir(parents=True, exist_ok=True)

        # Write to file
        tree.write(
            str(self.options.output_path),
            pretty_print=True,
            xml_declaration=True,
            encoding="UTF-8",
        )

        return self.options.output_path

    def _create_root(self) -> ET.Element:
        """Create root element with namespaces."""
        nsmap = {
            None: self.ARCHIMATE_NS,
            "xsi": self.XSI_NS,
        }

        root = ET.Element(
            "model",
            nsmap=nsmap,
            attrib={
                f"{{{self.XSI_NS}}}schemaLocation": f"{self.ARCHIMATE_NS} http://www.opengroup.org/xsd/archimate/3.1/archimate3_Diagram.xsd",
                "identifier": self._generate_id(),
                "version": "5.0.0",
            },
        )

        return root

    def _add_metadata(self, root: ET.Element) -> None:
        """Add model metadata."""
        metadata = ET.SubElement(root, "metadata")

        # Schema
        schema = ET.SubElement(metadata, "schema")
        schema.text = "3.2"

        # Creator
        creator = ET.SubElement(metadata, "creator")
        creator.text = "Documentation Robotics CLI"

        # Created
        created = ET.SubElement(metadata, "created")
        created.text = datetime.utcnow().isoformat() + "Z"

        # Name
        name = ET.SubElement(root, "name")
        name.text = self.model.manifest.project.get("name", "Architecture Model")

        # Documentation
        if "description" in self.model.manifest.project:
            documentation = ET.SubElement(root, "documentation")
            documentation.text = self.model.manifest.project["description"]

    def _add_elements(self, parent: ET.Element) -> None:
        """Add all elements to the export."""
        for layer in self.model.layers.values():
            # Skip non-ArchiMate layers
            if layer.name not in ["motivation", "business", "application", "technology"]:
                continue

            # Filter by layer if specified
            if self.options.layer_filter and layer.name != self.options.layer_filter:
                continue

            for element in layer.elements.values():
                self._add_element(parent, element)

    def _add_element(self, parent: ET.Element, element: Element) -> None:
        """Add a single element."""
        # Get ArchiMate type
        archimate_type = self.LAYER_TYPE_MAP.get(
            (element.layer, element.type), "ApplicationComponent"  # Default fallback
        )

        # Create element
        elem = ET.SubElement(
            parent,
            "element",
            attrib={
                "identifier": element.id,
                f"{{{self.XSI_NS}}}type": archimate_type,
            },
        )

        # Add name
        name = ET.SubElement(elem, "name")
        name.text = element.name

        # Add documentation
        if element.description:
            documentation = ET.SubElement(elem, "documentation")
            documentation.text = element.description

        # Add properties
        if self.options.include_metadata:
            self._add_properties(elem, element)

    def _add_properties(self, parent: ET.Element, element: Element) -> None:
        """Add element properties."""
        properties_section = ET.SubElement(parent, "properties")

        # Add all element data as properties (except standard fields)
        exclude_keys = {"id", "name", "description", "type"}

        for key, value in element.data.items():
            if key in exclude_keys:
                continue

            # Convert complex values to string
            if isinstance(value, (dict, list)):
                import json

                value_str = json.dumps(value)
            else:
                value_str = str(value)

            prop = ET.SubElement(
                properties_section, "property", attrib={"identifier": self._generate_id()}
            )

            prop_key = ET.SubElement(prop, "key")
            prop_key.text = key

            prop_value = ET.SubElement(prop, "value")
            prop_value.text = value_str

    def _add_relationships(self, parent: ET.Element) -> None:
        """Add all relationships to the export."""
        # Get all references from registry
        if not hasattr(self.model, "reference_registry"):
            return

        for ref in self.model.reference_registry.references:
            self._add_relationship(parent, ref)

    def _add_relationship(self, parent: ET.Element, reference) -> None:
        """Add a single relationship."""
        # Get ArchiMate relationship type
        archimate_type = self.RELATIONSHIP_TYPE_MAP.get(
            reference.reference_type, "Association"  # Default fallback
        )

        # Create relationship
        rel = ET.SubElement(
            parent,
            "relationship",
            attrib={
                "identifier": self._generate_id(),
                f"{{{self.XSI_NS}}}type": archimate_type,
                "source": reference.source_id,
                "target": reference.target_id,
            },
        )

        # Add name (optional)
        name = ET.SubElement(rel, "name")
        name.text = f"{reference.reference_type}: {reference.property_path}"

    def _add_organizations(self, parent: ET.Element) -> None:
        """Add organizational structure (folders/views)."""
        # Create folder for each layer
        for layer in self.model.layers.values():
            if layer.name not in ["motivation", "business", "application", "technology"]:
                continue

            folder = ET.SubElement(
                parent,
                "item",
                attrib={"identifier": self._generate_id(), "type": "folder"},
            )

            name = ET.SubElement(folder, "label")
            name.text = layer.config.get("name", layer.name.title())

            # Add elements to folder
            for element in layer.elements.values():
                item = ET.SubElement(folder, "item", attrib={"identifierRef": element.id})

    def _generate_id(self) -> str:
        """Generate unique identifier."""
        import uuid

        return f"id-{uuid.uuid4()}"

    def validate_output(self, output_path: Path) -> bool:
        """Validate exported ArchiMate XML."""
        try:
            # Parse XML
            tree = ET.parse(str(output_path))
            root = tree.getroot()

            # Check root element
            if root.tag != f"{{{self.ARCHIMATE_NS}}}model":
                return False

            # Define namespace for finding elements
            ns = {"archimate": self.ARCHIMATE_NS}

            # Check required sections
            required = ["metadata", "elements"]
            for section in required:
                if root.find(f"archimate:{section}", ns) is None:
                    return False

            return True

        except Exception as e:
            print(f"Validation error: {e}")
            return False

File: cli/src/export/archimate-exporter.ts
import type { Model } from "../core/model.js";
import type { Exporter, ExportOptions } from "./types.js";
import { escapeXml } from "./types.js";
import { isTelemetryEnabled, startSpan, endSpan } from "../telemetry/index.js";

/**
 * ArchiMate XML Exporter for layers 1, 2, 4, 5
 * Generates valid ArchiMate 3.2 XML format
 */
export class ArchiMateExporter implements Exporter {
  name = "ArchiMate";
  supportedLayers = ["motivation", "business", "application", "technology"];

  async export(model: Model, options: ExportOptions = {}): Promise<string> {
    const span = isTelemetryEnabled
      ? startSpan("export.format.archimate", {
          "export.layerCount": options.layers?.length || this.supportedLayers.length,
        })
      : null;

    try {
      const lines: string[] = [];

      // XML declaration
      lines.push('<?xml version="1.0" encoding="UTF-8"?>');

      // Root element with namespaces
      lines.push('<model xmlns="http://www.opengroup.org/xsd/archimate/3.0/"');
      lines.push('       xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"');
      lines.push(
        '       xsi:schemaLocation="http://www.opengroup.org/xsd/archimate/3.0/ http://www.opengroup.org/xsd/archimate/3.1/archimate3_Diagram.xsd">'
      );

      // Model metadata
      lines.push(`  <name>${escapeXml(model.manifest.name)}</name>`);
      if (model.manifest.description) {
        lines.push(`  <documentation>${escapeXml(model.manifest.description)}</documentation>`);
      }

      // Export elements - query graph model directly
      lines.push("  <elements>");

      const layersToExport = options.layers || this.supportedLayers;
      const nodesByLayer: Array<{ layer: string; node: any }> = [];

      // Query graph model for nodes in supported layers
      for (const layerName of layersToExport) {
        if (!this.supportedLayers.includes(layerName)) continue;

        const nodes = model.graph.getNodesByLayer(layerName);
        for (const node of nodes) {
          nodesByLayer.push({ layer: layerName, node });
        }
      }

      if (isTelemetryEnabled && span) {
        (span as any).setAttribute("export.elementCount", nodesByLayer.length);
      }

      for (const { layer, node } of nodesByLayer) {
        const archiType = this.mapToArchiMateType(node.type, layer);
        lines.push(`    <element identifier="${node.id}" xsi:type="${archiType}">`);
        lines.push(`      <name>${escapeXml(node.name)}</name>`);

        if (node.description) {
          lines.push(`      <documentation>${escapeXml(node.description)}</documentation>`);
        }

        // Add properties section
        lines.push("      <properties>");

        // Add node properties
        const propKeys = Object.keys(node.properties || {});
        if (propKeys.length > 0) {
          for (const key of propKeys) {
            const val = node.properties[key];

            // Special handling for source reference structure
            if (key === "source" && val && typeof val === "object" && "reference" in val) {
              const flattenedProps = this.flattenSourceReference(val.reference);
              for (const [flatKey, flatValue] of Object.entries(flattenedProps)) {
                lines.push(
                  `        <property key="${escapeXml(flatKey)}" value="${escapeXml(String(flatValue))}" />`
                );
              }
            } else {
              const strValue = this.valueToString(val);
              lines.push(
                `        <property key="${escapeXml(key)}" value="${escapeXml(strValue)}" />`
              );
            }
          }
        }

        lines.push("      </properties>");
        lines.push("    </element>");
      }

      lines.push("  </elements>");

      // Export relationships - query graph model edges
      lines.push("  <relationships>");

      let relationshipCount = 0;
      const nodeIds = new Set(nodesByLayer.map((n) => n.node.id));
      const edges = model.graph.getAllEdges();

      // Filter edges to include only those between exported nodes
      for (const edge of edges) {
        if (!nodeIds.has(edge.source) || !nodeIds.has(edge.destination)) {
          continue;
        }

        lines.push(`    <relationship identifier="${this.escapeId(edge.id)}" `);
        lines.push(`                  source="${edge.source}" target="${edge.destination}" `);
        lines.push(`                  xsi:type="Association">`);
        lines.push(`      <name>${edge.predicate}</name>`);
        lines.push(`    </relationship>`);
        relationshipCount++;
      }

      lines.push("  </relationships>");
      lines.push("</model>");

      const result = lines.join("\n");

      if (isTelemetryEnabled && span) {
        (span as any).setAttribute("export.relationshipCount", relationshipCount);
        (span as any).setAttribute("export.size", result.length);
        (span as any).setStatus({ code: 0 });
      }

      return result;
    } catch (error) {
      if (isTelemetryEnabled && span) {
        (span as any).recordException(error as Error);
        (span as any).setStatus({
          code: 2,
          message: error instanceof Error ? error.message : String(error),
        });
      }
      throw error;
    } finally {
      endSpan(span);
    }
  }

  /**
   * Map element type and layer to ArchiMate type
   */
  private mapToArchiMateType(type: string, layer: string): string {
    const mapping: Record<string, Record<string, string>> = {
      motivation: {
        stakeholder: "Stakeholder",
        driver: "Driver",
        assessment: "Assessment",
        goal: "Goal",
        outcome: "Outcome",
        principle: "Principle",
        requirement: "Requirement",
        constraint: "Constraint",
      },
      business: {
        "business-actor": "BusinessActor",
        "business-role": "BusinessRole",
        "business-collaboration": "BusinessCollaboration",
        "business-interface": "BusinessInterface",
        "business-process": "BusinessProcess",
        "business-function": "BusinessFunction",
        "business-interaction": "BusinessInteraction",
        "business-event": "BusinessEvent",
        "business-service": "BusinessService",
        "business-object": "BusinessObject",
      },
      application: {
        "application-component": "ApplicationComponent",
        "application-collaboration": "ApplicationCollaboration",
        "application-interface": "ApplicationInterface",
        "application-function": "ApplicationFunction",
        "application-interaction": "ApplicationInteraction",
        "application-process": "ApplicationProcess",
        "application-event": "ApplicationEvent",
        "application-service": "ApplicationService",
        "data-object": "DataObject",
      },
      technology: {
        node: "Node",
        device: "Device",
        "system-software": "SystemSoftware",
        "technology-collaboration": "TechnologyCollaboration",
        "technology-interface": "TechnologyInterface",
        path: "Path",
        "communication-network": "CommunicationNetwork",
        "technology-function": "TechnologyFunction",
        "technology-process": "TechnologyProcess",
        "technology-interaction": "TechnologyInteraction",
        "technology-event": "TechnologyEvent",
        "technology-service": "TechnologyService",
        artifact: "Artifact",
      },
    };

    return mapping[layer]?.[type] || "Element";
  }

  /**
   * Escape ID characters for use in identifiers
   */
  private escapeId(str: string): string {
    return str.replace(/[^a-zA-Z0-9_-]/g, "_");
  }

  /**
   * Convert value to string representation
   */
  private valueToString(value: unknown): string {
    if (typeof value === "string") return value;
    if (typeof value === "number") return String(value);
    if (typeof value === "boolean") return String(value);
    if (Array.isArray(value)) return `[${value.map((v) => this.valueToString(v)).join(", ")}]`;
    if (value && typeof value === "object") {
      return JSON.stringify(value);
    }
    return String(value);
  }

  /**
   * Flatten source reference structure into ArchiMate properties
   * Converts nested structure like { provenance: "extracted", locations: [...] }
   * into flat properties like "source.provenance", "source.file.0", etc.
   */
  private flattenSourceReference(sourceRef: any): Record<string, string> {
    const props: Record<string, string> = {};

    if (!sourceRef || typeof sourceRef !== "object") {
      return props;
    }

    // Add provenance
    if (sourceRef.provenance) {
      props["source.provenance"] = sourceRef.provenance;
    }

    // Add locations
    if (Array.isArray(sourceRef.locations)) {
      sourceRef.locations.forEach((location: any, index: number) => {
        if (location.file) {
          props[`source.file.${index}`] = location.file;
        }
        if (location.symbol) {
          props[`source.symbol.${index}`] = location.symbol;
        }
        if (location.line !== undefined) {
          props[`source.line.${index}`] = String(location.line);
        }
        if (location.column !== undefined) {
          props[`source.column.${index}`] = String(location.column);
        }
      });
    }

    // Add repository info
    if (sourceRef.repository) {
      if (sourceRef.repository.url) {
        props["source.repository.url"] = sourceRef.repository.url;
      }
      if (sourceRef.repository.commit) {
        props["source.repository.commit"] = sourceRef.repository.commit;
      }
      if (sourceRef.repository.branch) {
        props["source.repository.branch"] = sourceRef.repository.branch;
      }
    }

    return props;
  }
}
